### ISSUE TITLE
[Route] Route failure when using new in initializers in PHP 8.1

### ISSUE THREAD

**Issue by ytilotti** [symfony/symfony#48240]
### Symfony version(s) affected

6.1

### Description

PHP 8.1 introduces the ability to add [New in initializers](https://wiki.php.net/rfc/new_in_initializers). This method not working with routing system?

### How to reproduce

Run this route:

```php
<?php
declare(strict_types=1);

namespace App\Controller;

use App\Entity\AnEntity;
use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
use Symfony\Component\HttpFoundation\Request;
use Symfony\Component\HttpFoundation\Response;
use Symfony\Component\Routing\Annotation\Route;

class TestController extends AbstractController
{
    #[Route('/add/', name: 'controller_add')]
    #[Route('/edit/{id}/', name: 'controller_edit', requirements: ['id' => '\d+'])]
    public function editAction(Request $request, AnEntity $entity = new AnEntity()): Response
    {
        //...
    }
}
```

### Additional Context

```
App\Controller\TestController::editAction(): Argument # 2 ($entity) must be of type App\Entity\AnEntity, null given, called in /var/www/html/vendor/symfony/http-kernel/HttpKernel.php

  at src/Controller/TestController.php:64
  at App\Controller\TestController->editAction(object(Request), null)
     (vendor/symfony/http-kernel/HttpKernel.php:163)
  at Symfony\Component\HttpKernel\HttpKernel->handleRaw(object(Request), 1)
     (vendor/symfony/http-kernel/HttpKernel.php:75)
  at Symfony\Component\HttpKernel\HttpKernel->handle(object(Request), 1, true)
     (vendor/symfony/http-kernel/Kernel.php:202)
  at Symfony\Component\HttpKernel\Kernel->handle(object(Request))
     (public/index.php:25)     
```

**Comment by xabbuh** [comment]
Can you create a small example application that allows to reproduce your issue?

**Comment by ytilotti** [comment]
I created a small application on [phpsandbox](https://phpsandbox.io/e/x/roxnh?layout=EditorPreview&defaultPath=%2F&theme=dark&showExplorer=no&openedFiles=) to test that. Need to call [https://roxnh.ciroue.com/edit/1/](https://roxnh.ciroue.com/edit/1/) or [https://roxnh.ciroue.com/add/](https://roxnh.ciroue.com/add/).

**Comment by fancyweb** [comment]
I cannot see the all the code/files after opening your link but I guess you are using the `DoctrineParamConverter` from `sensio/framework-extra-bundle`. Since `$entity` has a default value, the param converter doesn't throw a `NotFoundHttpException` and returns `null`. But I think that even if it was returning the default value, it would not end up doing what you expect because `/edit/<id that doesn't exist>` wouldn't throw and gives you the default value instead. Param converters (and (argument) value resolvers) are "bound" to the controller method and not to the routes so they cannot be applied conditionally.

**Comment by ytilotti** [comment]
To access the files, you need to fork. Or the zip file: [roxnh.zip](https://github.com/symfony/symfony/files/10046350/roxnh.zip)

The error is different when `/edit/<id that doesn't exist>`. A classic ParamConvertor error, if I remove the `new`.

So, the error current on null is set before the ParamConvertor. Your point of view is interesting. But that is normal?
Doctrine bundle make a find so the error ParamConvertor will throw.

and why the same error of null on route `/add/`?

**Comment by wouterj** [comment]
Hi! This is not something related to Symfony, but to the [SensioFrameworkExtraBundle](https://github.com/sensiolabs/SensioFrameworkExtraBundle/). Given it will be deprecated in the next months (all features are merged into Symfony 6.2: https://symfony.com/blog/new-in-symfony-6-2-built-in-cache-security-template-and-doctrine-attributes ), I don't think this feature will be implemented there.

I also agree with @fancyweb that this is not something I would recommend using, if it worked as expected. I would set-up 2 controllers, one for adding new ones and one for editing, as both have different behavior (e.g. when trying to edit a non-existing ID, you probably don't want to add which is what this current controller would do).

If you (or anyone else) have any other use-case for using new initializers in a controller, please share them in this issue. For now, I'm going to close this one. Anyway, thanks for starting this discussion and sharing a reproducer :)

**Comment by ytilotti** [comment]
@wouterj, currently I have 2 controllers. `/add/` controller call the `/edit/{id}` controller.
But yesterday I was working on a migration from 6.0 to 6.1 and more precisely on the refactoring PHP 8.x. With the new features in PHP 8.1 this possibility seemed logical. So we reopen the issue on 6.2? :).

And yes I'm agree with you, if id not exists, `/add/` is called. Normal or not normal, thas is the question... because the pattern not match. An id not exists, need to throw an exception.
